### ISSUE TITLE
Application names longer than 37 characters cannot be executed from Application menu

### ISSUE THREAD

**Issue by ctr49** [QubesOS/qubes-issues#7766]
### Qubes OS release

Qubes release 4.1.1 (R4.1)

### Brief summary

qvm-sync-apps makes domU applications available in the "Application" menu in dom0.
Doing this it sync "desktop" file from various locations inside domU and makes them available for selection in dom0 by using Qubes Manager.
"Selected" applications will be available in ~/.local/share/applications, mimicking the original .desktop application much like:
`app.desktop` -> `qvm-run -q -a --service -- personal qubes.StartApp+app`
However, when  the filename "app" in "app.desktop" is longer than 37 characters, such an application menu item cannot be executed anymore.

This is apparently caused by a size limitation in qdb. qubes-read returns different error codes when an entry was not found ("2") or when there was another error "1". If the application name is longer than 37 characters, the string concatenation of "/qubes-service/app-dispvm." and the application name exceeds 64 characters, qdb_read returns a different error than "is_enoent", which makes /etc/qubes-rpc/qubes.StartApp throw an error.

### Steps to reproduce

using qubes.StartApp (in domU):
works fine with 37 characters:
```
user@personal:~$ /etc/qubes-rpc/qubes.StartApp 1234567890123456789012345678901234567
1234567890123456789012345678901234567 not found in /home/user/.local/share:/usr/local/share:/usr/share
```
but fails with 38 characters:
```
user@personal:~$ /etc/qubes-rpc/qubes.StartApp 12345678901234567890123456789012345678
Traceback (most recent call last):
  File "/etc/qubes-rpc/qubes.StartApp", line 39, in <module>
    main(sys.argv)
  File "/etc/qubes-rpc/qubes.StartApp", line 15, in main
    if qubes_db.read('/qubes-service/app-dispvm.' + arg) is not None:
qubesdb.Error: (0, 'Error')
```

The underlying reason for `qubesdb.Error` can also be seen when executing qubesdb-read with a 63 character string (including initial slash) in domU:
```
user@personal:~$ qubesdb-read /12345678901234567890123456789012345678901234567890123456789012 ; echo $?
Failed to read /12345678901234567890123456789012345678901234567890123456789012
2
```
and using a 64 character string (including initial slash):
```
user@personal:~$ qubesdb-read /123456789012345678901234567890123456789012345678901234567890123 ; echo $?
Failed to read /123456789012345678901234567890123456789012345678901234567890123
1
```

### Expected behavior

qubesdb-read returning consistent error codes regardless of query length / application being executed from Application menu for application names regardless of application name length

### Actual behavior

qubesdb-read returns error code "1" for queries longer than 63 characters / application not being executed from Application menu when application name longer than 37 characters

### Note

This is of special interest when running AppImage-based applications using appimaged. Appimaged creates an artificial application name "appimagekit_%Identifier%" which is *always* 44 characters long ("appimagekit_" + 32 byte identifier). Those applications can currently not be executed from the Applications menu at all.

**Comment by DemiMarie** [comment]
> ### Expected behavior
> 
> qubesdb-read returning consistent error codes regardless of query length / application being executed from Application menu for application names regardless of application name length

By design, `qubesdb-read` returns a different error code for “no such entry in QubesDB” and “cannot determine if an entry exists”.  This is so that callers can confirm the absense of an entry, not just the presence of one.  The real problem here is that qubesdb limits queries to 63 characters.

**Comment by marmarek** [comment]
The key names in qubesdb have limited length by design. This limit might get increased in the future, but I don't think it's necessary for fixing this issue. We can use not desktop name directly, but a hash of it (even md5 would work, intentional collisions are not really a concern here).

As a temporary workaround, we can skip `app-dispvm` feature check for longer desktop names _and_ refuse to set it in the first place (to avoid the case where it is set but isn't working).

**Comment by DemiMarie** [comment]
@marmarek: what about using `hash-dispvm.<SHA256 hash of name>`?  Also, setting an invalid service name should fail.

**Comment by marmarek** [comment]
SHA256 is 64 characters... but something like with another hash should work.

**Comment by DemiMarie** [comment]
> SHA256 is 64 characters... but something like with another hash should work.

Only if hex-encoded.  base64URL encoding is 44 bytes, so `/qubes-service/dvm.<base64URL SHA256 hash>` would work (with no bytes to spare).

**Comment by ctr49** [comment]
> > ### Expected behavior
> > qubesdb-read returning consistent error codes regardless of query length / application being executed from Application menu for application names regardless of application name length
> 
> By design, `qubesdb-read` returns a different error code for “no such entry in QubesDB” and “cannot determine if an entry exists”. This is so that callers can confirm the absense of an entry, not just the presence of one. The real problem here is that qubesdb limits queries to 63 characters.

I agree about the real problem in this context (and also the suggested workaround) and also to the intention that `qubesdb-read` can distinguish between an error and something not being found. However, if there is a size limitation, it it obvious that a request that is larger than this size limitation cannot be found, isn't it?
So in parallel to the long-term size adjustment of qdb and the short-term usage of base64-encoded hashes instead of the original app file name, wouldn't it be an option to return something more meaningful from qubes-read (i.e. requested size larger than length limit) to allow consumers of qubesdb-read handle this situation gracefully? After all, even if you increase to 128, 256 or even 1k, the request may still be larger than the size limit and qubesdb-read is used in many places.

**Comment by andrewdavidwong** [comment]
Duplicate of #7760

**Comment by andrewdavidwong** [comment]
This appears to be a [duplicate](https://www.qubes-os.org/doc/issue-tracking/#new-issues-should-not-be-duplicates-of-existing-issues) of an existing issue. If so, please comment on the appropriate existing issue instead. If anyone believes this is not really a duplicate, please leave a comment briefly explaining why. We'll be happy to take another look and, if appropriate, reopen this issue. Thank you.
